### PR TITLE
[V0][Fix] structured decoding compatibility with speculative decoding

### PR DESCRIPTION
**Pull request by southfreebird**
This PR was created by the Nebius team.

The main focus of this PR is to fix guided generation for speculative decoding. We found that when using the xGrammar backend with speculative decoding, vLLM crashes [here](https://github.com/southfreebird/vllm/blob/main/vllm/model_executor/guided_decoding/xgrammar_decoding.py#L321). This PR addresses the issue by using a rollback mechanism in xGrammar.

### CODE REVIEW

**Comment by hmellor** [comment]
Thanks for the contribution!

Please can you make sure that your commits are signed off (instructions [here](https://github.com/vllm-project/vllm/pull/13823/checks?check_run_id=37776376917)).

Also, some of the entrypoint tests are failing with:
```
[2025-02-25T12:06:05Z]     def _vocab_size(self) -> int:
[2025-02-25T12:06:05Z]         """Get the vocab size of the model and make sure it's consistent between
[2025-02-25T12:06:05Z]         draft and target workers.
[2025-02-25T12:06:05Z]         """
[2025-02-25T12:06:05Z]         vocab_sizes = [
[2025-02-25T12:06:05Z]             worker.vocab_size
[2025-02-25T12:06:05Z]             for worker in [self.proposer_worker, self.scorer_worker]
[2025-02-25T12:06:05Z]         ]
[2025-02-25T12:06:05Z] >       assert all(vocab_sizes[0] == vocab_size for vocab_size in vocab_sizes)
[2025-02-25T12:06:05Z] E       AssertionError
```
Which appears to be relevant.

**Comment by mgoin** [comment]
Oh this is interesting, I did not consider the need for rollback until this case. Thanks for your work. I think it is crucial to add a test using guided decoding and speculation together since AFAIK we haven't used these together

**Comment by mgoin** [comment]
@russellb @aarnphm

**Comment by aarnphm** [comment]
I will look into this td for the v1. But thanks for making the PR.

File: tests/entrypoints/llm/test_guided_generate.py [hunk 1]
```diff
@@ -12,21 +12,26 @@
 from vllm.outputs import RequestOutput
 from vllm.sampling_params import GuidedDecodingParams, SamplingParams
 
-MODEL_NAME = "Qwen/Qwen2.5-1.5B-Instruct"
```
Comment (aarnphm) [inline tests/entrypoints/llm/test_guided_generate.py:15]: can you keep this mdoel? is there a specific reason for using a larger models for ci?

File: tests/entrypoints/llm/test_guided_generate.py [hunk 2]
```diff
@@ -12,21 +12,26 @@
...
+        elif mode == "speculative":
+            llm_kwargs = {
+                # the model with fixed vocabulary size
+                "speculative_model": "tugstugi/Qwen2.5-Coder-0.5B-QwQ-draft",
+                "num_speculative_tokens": 3,
+            }
+        else:
+            raise ValueError(f"Unsupported LLM mode: {mode}")
+
+        return llm_kwargs
```
Comment (aarnphm) [inline tests/entrypoints/llm/test_guided_generate.py:34]: Let's try to reduce the amount of change that is not relevant to the PR.

I don't see the diff here. Given that it is a pytest fixture, there is no need to raise exception here.

if the test uses wrong, params, it won't even run the test.

File: tests/entrypoints/llm/test_guided_generate.py (hunk 1 above)
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:15]: Sorry, my bad.
I thought the 7B model was used initially. Reverting to the 1.5B.

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 3]
```diff
@@ -306,18 +318,28 @@ def __call__(self, input_ids: list[int],
             self._ensure_ctx()
 
         if len(self.matchers) == 0:
+            max_rollback_tokens = (self.config.num_lookahead_slots +
```
Comment (aarnphm) [inline vllm/model_executor/guided_decoding/xgrammar_decoding.py:321]: qq: why do we need to append 1 to the num_lookahead_slots here?

Comment (southfreebird) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py]: I double-checked this place; it's safe to remove this +1
Initially, it was a workaround for the bonus token

File: tests/entrypoints/llm/test_guided_generate.py [hunk 4]
```diff
@@ -16,11 +16,22 @@
...
+def llm(request):
+
+    def get_llm_kwargs(mode: str):
+        if mode == "autoregressive":
+            return {}
+        return {
+            # the model with fixed vocabulary size
+            "speculative_model": "Qwen/Qwen2.5-0.5B-Instruct",
+            "num_speculative_tokens": 3,
+        }
```
Comment (mgoin) [inline tests/entrypoints/llm/test_guided_generate.py:29]: change this to be an `elif mode == "speculative":` and raise an exception in the else case saying unsupported mode

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 3 above)
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py]: sg thanks for the explanation.

File: tests/entrypoints/llm/test_guided_generate.py [hunk 5]
```diff
@@ -16,11 +16,22 @@
...
+
+    test_llm_kwargs = get_llm_kwargs(request.param)
     # pytest caches the fixture so we use weakref.proxy to
     # enable garbage collection
-    llm = LLM(model=MODEL_NAME, max_model_len=1024)
+    llm = LLM(model=MODEL_NAME, max_model_len=1024, **test_llm_kwargs)
```
Comment (mgoin) [inline tests/entrypoints/llm/test_guided_generate.py:34]: I'm not sure I understand how this works - will this fixture now run all of the tests in this file for each entry in `params`?

File: tests/entrypoints/llm/test_guided_generate.py (hunk 4 above)
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:29]: The opposite suggestion in the comment: https://github.com/vllm-project/vllm/pull/13823#discussion_r1969826373
:)

File: tests/entrypoints/llm/test_guided_generate.py (hunk 5 above)
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:34]: It first runs all tests with "autoregressive" and then loads the "speculative" model and runs all the tests in the file with it

Comment (aarnphm) [reply tests/entrypoints/llm/test_guided_generate.py:34]: actually for pytest fixture this would run both cases, so no need for exception

```python

def test_use_llms(llm):
    ...
```

then there would be two tests `test_use_llms[llm_autoregressive]` and `test_use_llms[llm_speculative]`

https://docs.pytest.org/en/6.2.x/fixture.html#parametrizing-fixtures

**Comment by southfreebird** [comment]
Hi @mgoin @aarnphm
I rebased the PR and resolved the conflicts.

Do I need to do anything else on my end? Are you waiting for me to make any changes to the code?
I'm not sure if I need to fix this: https://github.com/vllm-project/vllm/pull/13823#discussion_r1970315741

**Comment by aarnphm** [comment]
hmm it seems like the test failure is not related?

**Comment by aarnphm** [comment]
Can you rename the PR title accordingly? 

to `[V0][Fix] structured decoding compatibility with speculative decoding`

**Review by aarnphm — APPROVED** [review]
seems like the tests failure is not related, but the tests for this passes so LGTM.

**Comment by aarnphm** [comment]
@benchislett you might be interested in this PR

**Review by benchislett — CHANGES_REQUESTED** [review]
I have concerns about this PR in its current state. It seems incompatible with a number of core features (ngram speculative decoding, MQA scoring/enforce_eager), and the correctness of the rollback logic seems fragile.

File: vllm/spec_decode/draft_model_runner.py [hunk 6]
```diff
@@ -71,6 +71,12 @@ def _update_sampling_metadata(self, sampling_metadata, num_seqs,
             assert seq_group.prompt_logprob_indices == []  # No prompt
             assert seq_group.sample_indices == [i]  # Simple
 
+            for seq_id in seq_group.seq_ids:
```
[the remark this answers is not in the document]
Comment (southfreebird) [reply vllm/spec_decode/draft_model_runner.py:76]: This is valid
Fixed it by applying only to requests with guided decoding enabled

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 7]
```diff
@@ -335,18 +347,28 @@ def __call__(self, input_ids: list[int],
...
-        if not self.prefilled:
-            # Have not sampled a token yet
-            self.prefilled = True
-        else:
+        for i in range(len(self.matchers)):
+            if self.num_processed_tokens[i] > 0 and self.num_processed_tokens[
+                    i] >= len(input_ids):
+                diff = self.num_processed_tokens[i] - len(input_ids) + 1
+                self.num_processed_tokens[i] -= diff
+                self.matchers[i].rollback(diff)
```
[the remark this answers is not in the document]
Comment (southfreebird) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:378]: @benchislett 
Could you please help me find any examples of n-gram speculative decoding? I'd be great to have a test or some reference to check this

File: tests/entrypoints/llm/test_guided_generate.py [hunk 8]
```diff
@@ -17,11 +17,22 @@
 GUIDED_DECODING_BACKENDS = ["outlines", "lm-format-enforcer", "xgrammar"]
 
 
-@pytest.fixture(scope="module")
-def llm():
+@pytest.fixture(scope="module", params=["autoregressive", "speculative"])
+def llm(request):
```
[the remark this answers is not in the document]
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:21]: It seems stochastic, but it works for me with seed=0

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:378]: Here's an example from the docs. 
https://docs.vllm.ai/en/latest/features/spec_decode.html#speculating-by-matching-n-grams-in-the-prompt

File: tests/entrypoints/llm/test_guided_generate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:21]: Agree.
The problem is that it can fail to stop generating output after "}", regardless of whether speculative decoding is used or not. This issue is fixed in main by specifying the seed, as I understand.

My point is that it doesn’t seem to be a problem with speculative decoding itself. However, it is unexpected behaviour for me

[the remark this answers is not in the document]
Comment (benchislett) [reply tests/entrypoints/llm/test_guided_generate.py:21]: It does not seem clear to me that this is what is happening. To my understanding, when the matcher reaches a terminal state it will restrict the next token to be the end-of-sequence token, terminating the completion. Why would it fail to stop generating output?

The output from my above error report does not look like this is what happened:
```
{"clarifying_question": "", "cost_per_serving": ">$5", "calories": "}} Can you provide more context or clarify your q...onTemArray<MultipartUploadRequest<TemData,Versions.MLVersionInfo,urlMap>", "kitchen_ids": "$nil", "holiday_ids": "{}"}
```
This looks like a guidance failure, where the matcher state became out-of-sync with the completion resulting in incorrect outputs.

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (southfreebird) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:378]: > This code seems to be incompatible with the MQA scorer, which is used when enforce-eager is True. I'm not sure if a small tweak could make this compatible or if a larger rework is required.

I investigated the problem with the MQA scorer and found that the main issue is that when the MQA scorer is chosen, the `LogitsProcessor` is not called on every verified token but only on the last one.

To clarify, when we call `BatchExpansionTop1Scorer.score_proposals`, it triggers `LogitsProcessor` for every proposed token.

BatchExpansionTop1Scorer:
```
Prefill of target:
input_ids: ()
Profile of draft:
input_ids: ()
Process logits for 2 proposal tokens:
input_ids: (90)
input_ids: (90, 1)  
Verify logits for 2 proposal and 1 bonus token:  
input_ids: (90)  
input_ids: (90, 15)  
input_ids: (90, 15, 30)
```
Works fine. ^^^

MQA Scorer:
```
Prefill of target:
input_ids: ()
Profile of draft:
input_ids: ()
Process logits for 2 proposal tokens:
input_ids: (90)
input_ids: (90, 1)  
Verify logits only for 1 bonus token:  
input_ids: (90, 65, 76)
```
Crashes on the next iteration. ^^^

It seems that the MQA scorer does not properly work with guided output. I suggest throwing an error when guided output is enabled and the MQA scorer is selected for now, and then fixing it further, as it looks like a significant amount of work (I can help with that if you don't mind).

What do you think?

[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:378]: Yes, a modification would need to be made to the MQA scorer to call the logits processor on the logits at each sample position.

I'll leave it up to the core reviewers to decide if this needs to be included or if it can be left for a later work

[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:378]: cc @WoosukKwon on plans for speculative decoding in v0. I think this is still important given that spec decode in v1 is still considered experimental.

[the remark this answers is not in the document]
Comment (southfreebird) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:370]: I checked NGram decoding, and the main challenge is that the draft tokens produced by NGram are not aligned with the guided output. As a result, they are not accepted, which leads to poor performance. I decided to disable them when guided output is enabled.

I also tried to fix this issue, but it resulted in very complex logic in XGrammarLogitsProcessor and outlines processors. Therefore, I suggest turning it off, as it ensures correct logic and outputs.

File: tests/entrypoints/llm/test_guided_generate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:23]: I also checked the tests again, and it seems that guided output works correctly. The problem lies in the special tokens (e.g., `\t` in this case) that the model can place inside fields. It seems that `xgrammar` does not check them.

I also found that a model without speculative decoding can generate them as well, so I suppose this is not an issue with this PR.

**Review by aarnphm — COMMENTED** [review]
Want to hear more your thoughts on this:

I talked with Woosuk offline, and I think that to make structured outputs properly work with spec decode in v0 will requires cover the cardinality of a small subsets of backends -> spec decode worker (mqa, eagle, draft), which requires a significant amount of work.

Given that we are focusing on moving to v1 soon, I'm thinking if it is better to focus all of the effort there, while in v0 we can say that "structured outputs won't work with spec decode"

I also am not familiar with the spec decode perf in v0, so I don't have much saying here (as I haven't explore spec decode deeply in v0)

cc @benchislett on this

**Comment by southfreebird** [comment]
We are using spec_decoding in v0, and we really need this feature because there are still many changes around it, such as the new guided backend.
This PR fixes a lot of problems in v0 and is very helpful for us. Since this PR resolves many issues and leads to correct behaviour rather than causing a core dump in vLLM, I don’t really understand the motivation for not adding it upstream.
Focusing on v1 might be a good idea, but since it doesn’t fully support spec_decoding, we still want to use spec_decoding in v0.

**Comment by aarnphm** [comment]
I'm also good with supporting a small subsets of spec decode features and clearly stated which one works with structured outputs.

**Comment by southfreebird** [comment]
Based on that, does this mean we're good to merge, or do you want something alse from my end?

File: tests/entrypoints/llm/test_guided_generate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (benchislett) [reply tests/entrypoints/llm/test_guided_generate.py:23]: I don't understand what you mean. Does this imply that there is a bug with xgrammar?

[the remark this answers is not in the document]
Comment (southfreebird) [reply tests/entrypoints/llm/test_guided_generate.py:23]: If the generation of special symbols inside a JSON field (which can be loaded with strict=False) is a bug then yes

**Comment by southfreebird** [comment]
Hi team,
I just wanted to check if you’re not planning to merge this PR. Thanks!

**Comment by ItzAmirreza** [comment]
Hi team :)
Any updates?

**Comment by russellb** [comment]
First of all, thank you very much for your hard work and contribution!

I thought about this and discussed it with some other maintainers; we decided it would be best not to merge this. The reasons are roughly:

* This is fixed in V1 via #14702 
* This is a relatively risky change to core code, and we're trying to spend as little time as possible on V0
* We feel time is better spent continuing the development on the V1 side than risking any accidental regressions in V0

Folks are welcome to continue using this in custom builds in the meantime, but I hope it won't be too long before everything needed is supported in V1.

Thank you again for the hard work, and I apologize that the PR has been in limbo for this long.
